inferno-core/src/Inferno/Instances/Arbitrary.hs

```diff
@@ -0,0 +1,629 @@
+          <$> ( oneof
+                  $ map pure
+                  $ concatMap
+                    ( \(prec, xs) ->
+                        concatMap
+                          ( \case
+                              (PrefixOp, _, op) -> [(prec, op)]
+                              _ -> []
+                          )
+                          xs
+                    )
+                  $ IntMap.toList baseOpsTable
+              )
+          <*> (arbitrarySized $ n `div` 3)
+          <*> arbitrary
+          <*> arbitrary
+      
+      -- FIXME: Currently broken, if we include this one we hang parseExpr test
+      arbitraryCase n =
+        (\e cs p1 p2 p3 -> Case p1 e p2 (NonEmpty.fromList cs) p3)
+          <$> (arbitrarySized $ n `div` 3)
+          <*> do
+            k <- choose (0, n)
+            sequence
+              [ (\i e p1 p2 -> (p1, i, p2, e))
+                  <$> arbitrarySizedPat (n `div` 3)
+                  <*> arbitrarySized (n `div` 3)
+                  <*> arbitrary
+                  <*> arbitrary
+                | _ <- [1 .. k]
+              ]
+            `suchThat` (not . null)
+          <*> arbitrary
+          <*> arbitrary
+          <*> arbitrary
+
+      arbitraryBracketed n =
+        Bracketed <$> arbitrary <*> arbitrarySized (n `div` 3) <*> arbitrary
+
+      arbitraryArrayComp n =
+        ArrayComp
+          <$> arbitrary
+          <*> (arbitrarySized $ n `div` 3)
+          <*> arbitrary
+          <*> ( NonEmpty.fromList
+                  <$> do
+                    k <- choose (0, n)
+                    sequence [(,,,,) <$> arbitrary <*> arbitrary <*> arbitrary <*> arbitrarySized (n `div` 3) <*> pure Nothing | _ <- [1 .. k]]
+                    `suchThat` (not . null)
+              )
+          <*> oneof [(\p e -> Just (p, e)) <$> arbitrary <*> (arbitrarySized $ n `div` 3), pure Nothing]
+          <*> arbitrary
+
+      arbitrarySized :: (Arbitrary hash, Arbitrary pos) => Int -> Gen (Expr hash pos)
+      arbitrarySized 0 =
+        oneof
+          [ arbitraryVar,
+            arbitraryEnum,
+            arbitraryLit,
+            Empty <$> arbitrary
+          ]
+      arbitrarySized n =
+        oneof
+          [ arbitraryVar,
+            arbitraryEnum,
+            arbitraryLit,
+            arbitraryApp n,
+            arbitraryLam n,
+            arbitraryLet n,
+            arbitraryIString n,
+            arbitraryIf n,
+            arbitraryOp n,
+            arbitraryPreOp n,
+            Array
+              <$> arbitrary
+              <*> ( do
+                      k <- choose (0, n)
+                      sequence [(,Nothing) <$> arbitrarySized (n `div` 3) | _ <- [1 .. k]]
+                  )
+              <*> arbitrary,
+            arbitraryArrayComp n,
+            One <$> arbitrary <*> arbitrarySized (n `div` 3),
+            Empty <$> arbitrary,
+            arbitraryAssert n,
+            -- arbitraryCase n -- Broken!
```

Comment:
**FIXME: test: "parseExpr and pretty are inverse up to normalizeExpr" hangs if we use them!**

Reply:
done!